### ISSUE TITLE
reading from Q6L returns variable values even if drive is off

### ISSUE THREAD

**Issue by peterferrie** [AppleWin/AppleWin#303]
Lady Tut turns off the drive, waits far more than 1s, and then periodically reads from Q6L, expecting the value to be constant.
908B                 LDA     $C088,X    ;turn off drive 
 908E loc_908E: 
 908E                 LDA     $C08C,X 
 9091                 PHA 
 9092                 LDY     #$10  ;approaching 3s
 9094 loc_9094: 
 9094                 LDA     #0 
 9096                 JSR     $FCA8 
 9099                 DEY 
 909A                 BNE     $9094 
 909C                 PLA 
 909D                 CMP     $C08C,X 
 90A0                 BNE     $908E ;wait for 2 sequential matching values
 90A2                 STX     $7F 
 ...
 0405                 STA     $7E    ;save the last read value  
 ...
 6000                 LDY     $7F    ;#$60 for slot 6 
 6002                 LDA     ($98),Y    ;this is $C08C, Y 
 6004                 CMP     $7E    ;compare with last read byte 
 6006                 BEQ     $600B 
 6008                 JMP     ($9A)  ;reboot instead

**Comment by TommyGH** [comment]
This description seems identical to a problem (and fix) I reported about 2 years ago with the disk drive latch values changing after the disk as stopped "spinning".
FYI,
Tom Ewers (Tommy)

**Comment by peterferrie** [comment]
> This description seems identical to a problem (and fix) I reported about 2
> years ago with the disk drive latch values changing after the disk as
> stopped "spinning".

If you can find the issue number, and verify that it's the same thing,
then we can close this one.
It all rests on how long exists between the motor off instruction and
the polling.
However, the issue exists in 1.25.03, so it's not fixed.

**Comment by sicklittlemonkey** [comment]
Sure, I know for a fact this won't work as is. We only tend to make emulation more accurate when there is a case for doing so. Unlike MESS. ; - )

Then again, this borders on protection that should really be removed ... I mean, there's no reason for software to do this - not even for protection. This only protects against running on an insufficiently accurate emulator.

Tommy, I can't find any issue you've logged that's similar.

Cheers,
Nick.

**Comment by peterferrie** [comment]
> Sure, I know for a fact this won't work as is. We only tend to make
> emulation more accurate when there is a case for doing so. Unlike MESS. ; -
> )

:-)

> Then again, this borders on protection that should really be removed ... I
> mean, there's no reason for software to do this - not even for protection.
> This only protects against running on an insufficiently accurate emulator.

And anyone debugging on real hardware, who steps over the routine
because it takes too long to run.
It's definitely a difficult problem - on the one hand, you have Mr
Do!, Randamn, and Sherwood Forest (at least) that race the drive to
read the sectors after turning off the drive; and on the other hand,
you have Lady Tut which intentionally waits for the drive to turn off,
and then verifies that it's really off.

**Comment by sicklittlemonkey** [comment]
> It's definitely a difficult problem - on the one hand, you have Mr
> Do!, Randamn, and Sherwood Forest (at least) that race the drive to
> read the sectors after turning off the drive; and on the other hand,
> you have Lady Tut which intentionally waits for the drive to turn off,
> and then verifies that it's really off.

Yeah, I added more accurate latch emulation for reading with the drive off
to AppleWin and a few other emulators because of Sherwood Forest etc.

This takes it a step further. Does LT do this once or often? Is it actually
trying to achieve something - like silence so it can play some music?

**Comment by peterferrie** [comment]
> Yeah, I added more accurate latch emulation for reading with the drive off
> to AppleWin and a few other emulators because of Sherwood Forest etc.
>
> This takes it a step further. Does LT do this once or often? Is it actually
> trying to achieve something - like silence so it can play some music?

LT checks periodically that the drive is still off, as a
copy-protection enforcement.  If the drive is found to be on, then LT
reboots immediately.

**Comment by TommyGH** [comment]
You likely don't need this information, but I thought I'd explain what I was referring to in my previous post.

Around Feb 2012, I was working on de-compiling WizardryIII and needed an AppleWin with 4 disk drives.  I modified AppleWin 1.20.1.0 to have 4 disk drives and ran into the existing AppleWin bug (while booting Pascal) with the drive latch changing after the disk drive had stopped spinning.  This caused the Pascal boot to hang.

I believe I reported this on the Berlios system, and if not I at least sent an e-mail to TomCh.

I also documented this bug in the Usenet thread in comp.sys.apple2:

  Re-engineered: Wizardy III, Legacy of Llylgamyn

See the entry for 24 May 2012, titled "AppleWin and 4 Disk Drives".

In that post, I included a short test program that demonstrated different behavior in the AppleWin emulator compared to the actual hardware.  I tested this on an AppleII+ or possibly a IIe.

Using Apple Pascal with AppleWin and 2 disk drives (on one controller) will never encounter the problem because the code that checks for the disk drive to really be stopped spinning is only checked when switching from a disk on one controller to a disk on another controller.

I identified the problem was in the routine:  DiskReadWrite.

If you need anything else, feel free to ask.
  
Regards,
Tommy

**Comment by sicklittlemonkey** [comment]
Right, I knew you had contacted Tom about the 4 drive request, but it looks like he took the discussion to email, as the GitHub Issue mentions the request but not the problem you encountered:
https://github.com/AppleWin/AppleWin/issues/141

The c.s.a2 discussion is here, but I didn't read the relevant part of the thread until now:
https://groups.google.com/d/topic/comp.sys.apple2/2oDJTbQaJWU/discussion

Just as a note, this drive spinning emulation is as old the hills. Back in 1994 or so I was working with a friend on an Apple II emulator in ARM assembly, and discovered RWTS's SAMESLOT routine. This does the same thing - compare two latch reads in succession to check the drive is spinning. The slowdown is huge unless you fake the spinning, so my approach was just to return 0x7F for one in every 16 nibbles.

Anyway, I'll try to get this issue sorted soon. Unfortunately having AppleWin's slots configurable for multiple drive controllers etc is still not on the roadmap. I hope your awesome Wizardry re-engineering continues despite the bugs and quirks of AppleWin. (That recent 128k performance issue is another one that's not trivial so solve.)

Cheers,
Nick.

**Comment by peterferrie** [comment]
> Just as a note, this drive spinning emulation is as old the hills. Back in
> 1994 or so I was working with a friend on an Apple II emulator in ARM
> assembly, and discovered RWTS's SAMESLOT routine. This does the same thing -
> compare two latch reads in succession to check the drive is spinning. The
> slowdown is huge unless you fake the spinning, so my approach was just to
> return 0x7F for one in every 16 nibbles.

Maybe you can check the number of elapsed cycles since the motor off
instruction, and if it's larger than N, just return a constant value?
There are probably more elegant ways, of course.

**Comment by sicklittlemonkey** [comment]
Sure, in AppleWin we already count cycles so it will be easy.

On an ARM back in 1994 we didn't have the cycles to count cycles, so I've
usually used a latch read counter & 0x0F to wrap. ; - )
​
Cheers,
Nick.
